File: backend/app/models/media.py
from datetime import datetime

from pydantic import BaseModel


class Server(BaseModel):
    """A Plex server belonging to the user."""

    name: str
    address: str
    port: int
    scheme: str = "http"
    local: bool = False
    owned: bool = True
    client_identifier: str

    @property
    def url(self) -> str:
        return f"{self.scheme}://{self.address}:{self.port}"


class Library(BaseModel):
    """A library section within a Plex server."""

    key: str
    title: str
    type: str  # 'movie', 'show', 'artist', 'photo'
    agent: str | None = None
    scanner: str | None = None
    thumb: str | None = None
    count: int | None = None


class MediaItem(BaseModel):
    """A movie or TV show from a Plex library."""

    rating_key: str
    guid: str  # Universal identifier for matching with watchlist
    title: str
    type: str  # 'movie' or 'show'
    summary: str | None = None
    year: int | None = None
    thumb: str | None = None
    art: str | None = None
    duration_ms: int | None = None
    added_at: datetime | None = None
    originally_available_at: str | None = None

    # For AI recommendations
    genres: list[str] = []
    rating: float | None = None
    content_rating: str | None = None
    view_count: int | None = None
    last_viewed_at: datetime | None = None

    # TV show specific
    season_count: int | None = None
    episode_count: int | None = None


class PaginatedResponse(BaseModel):
    """Paginated list of media items."""

    items: list[MediaItem]
    total: int
    offset: int
    limit: int
    has_more: bool


class WatchlistStatus(BaseModel):
    """Watchlist status for a media item."""

    rating_key: str
    title: str
    on_watchlist: bool


class WatchlistItem(BaseModel):
    """An item on the user's watchlist."""

    guid: str  # Universal identifier (e.g., plex://movie/...)
    title: str
    type: str
    year: int | None = None
    thumb: str | None = None

File: frontend/src/api/media.ts
import { apiFetch } from './client';

export interface Server {
  name: string;
  address: string;
  port: number;
  scheme: string;
  local: boolean;
  owned: boolean;
  client_identifier: string;
}

export interface Library {
  key: string;
  title: string;
  type: string;
  agent: string | null;
  scanner: string | null;
  thumb: string | null;
  count: number | null;
}

export interface MediaItem {
  rating_key: string;
  guid: string;
  title: string;
  type: string;
  summary: string | null;
  year: number | null;
  thumb: string | null;
  art: string | null;
  duration_ms: number | null;
  added_at: string | null;
  originally_available_at: string | null;
  genres: string[];
  rating: number | null;
  content_rating: string | null;
  view_count: number | null;
  last_viewed_at: string | null;
  season_count: number | null;
  episode_count: number | null;
}

export interface PaginatedResponse {
  items: MediaItem[];
  total: number;
  offset: number;
  limit: number;
  has_more: boolean;
}

export function getThumbnailUrl(serverName: string, thumbPath: string): string {
  const token = localStorage.getItem('auth_token');
  return `http://localhost:8000/api/media/thumbnail?server_name=${encodeURIComponent(serverName)}&path=${encodeURIComponent(thumbPath)}&token=${encodeURIComponent(token || '')}`;
}

export async function getServers(): Promise<Server[]> {
  return apiFetch<Server[]>('/api/media/servers');
}

export async function getLibraries(serverName: string): Promise<Library[]> {
  return apiFetch<Library[]>(
    `/api/media/libraries?server_name=${encodeURIComponent(serverName)}`
  );
}

export async function getLibraryItems(
  serverName: string,
  libraryKey: string,
  offset: number = 0,
  limit: number = 50
): Promise<PaginatedResponse> {
  return apiFetch<PaginatedResponse>(
    `/api/media/libraries/${libraryKey}/items?server_name=${encodeURIComponent(serverName)}&offset=${offset}&limit=${limit}`
  );
}

export async function clearCache(): Promise<{ message: string }> {
  return apiFetch<{ message: string }>('/api/media/cache', {
    method: 'DELETE',
  });
}

export interface WatchlistStatus {
  rating_key: string;
  title: string;
  on_watchlist: boolean;
}

export interface WatchlistItem {
  guid: string;
  title: string;
  type: string;
  year: number | null;
  thumb: string | null;
}

export async function getWatchlist(): Promise<WatchlistItem[]> {
  return apiFetch<WatchlistItem[]>('/api/media/watchlist');
}

export async function getWatchlistStatus(
  serverName: string,
  ratingKey: string
): Promise<WatchlistStatus> {
  return apiFetch<WatchlistStatus>(
    `/api/media/watchlist/status?server_name=${encodeURIComponent(serverName)}&rating_key=${encodeURIComponent(ratingKey)}`
  );
}

export async function addToWatchlist(
  serverName: string,
  ratingKey: string
): Promise<WatchlistStatus> {
  return apiFetch<WatchlistStatus>(
    `/api/media/watchlist?server_name=${encodeURIComponent(serverName)}&rating_key=${encodeURIComponent(ratingKey)}`,
    { method: 'POST' }
  );
}

export async function removeFromWatchlist(
  serverName: string,
  ratingKey: string
): Promise<WatchlistStatus> {
  return apiFetch<WatchlistStatus>(
    `/api/media/watchlist?server_name=${encodeURIComponent(serverName)}&rating_key=${encodeURIComponent(ratingKey)}`,
    { method: 'DELETE' }
  );
}
